### PR TITLE
DeepSpeed Comm. Backend v1

### PR DESCRIPTION
**Pull request by awan-10**
This PR introduces the DeepSpeed Comm. Backend (v1). 

Current advanced communication schemes rely on mixing python-level communication packages (e.g. torch.distributed, mpi4py for 1-bit Adam). In order to simplify comms prototypes, we're looking to add support for custom communication backends within DeepSpeed built directly on top of their respective libraries (e.g. NCCL, MPI, etc). 

This PR completes the first phase towards this goal by introducing:

- The new comms interface `deepspeed.comms`
- A complete wrapper around `torch.distributed` called `TorchBackend` for backwards-compatibility
- A rough skeleton for custom backends that we can use for phase 2

Co-authored-by: Quentin Anthony <qganthony@yahoo.com>
Co-authored-by: Ammar Ahmad Awan <ammar.awan@microsoft.com>
Co-authored-by: Jeff Rasley <jerasley@microsoft.com>

### CODE REVIEW

File: deepspeed/__init__.py [hunk 1]
```diff
@@ -109,11 +109,11 @@ def initialize(args=None,
         * ``lr_scheduler``: Wrapped lr scheduler if user ``lr_scheduler`` is passed, or
           if ``lr_scheduler`` specified in JSON configuration. Otherwise ``None``.
     """
-    log_dist("DeepSpeed info: version={}, git-hash={}, git-branch={}".format(
+    logger.info("DeepSpeed info: version={}, git-hash={}, git-branch={}".format(
```
Comment (awan-10) [inline deepspeed/__init__.py:112]: revert to master and test.

File: deepspeed/comm/backend.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
+''' DeepSpeed Communication Backend
+
+# DS Backends -- Direct C/Ops
```
Comment (awan-10) [inline deepspeed/comm/backend.py:3]: revise/read this documentation and fix.

File: deepspeed/moe/sharded_moe.py [hunk 3]
```diff
@@ -80,11 +80,17 @@ def gumbel_rsample(shape: Tuple, device: torch.device) -> Tensor:
...
+# einsum dimensions: (g)roup, (s)equence, (e)xpert, (m)odel, (c)apacity
+# See https://arxiv.org/pdf/2006.16668.pdf for details.
+
+
 # Based on https://github.com/pytorch/pytorch/pull/40762
 class _AllToAll(torch.autograd.Function):
     @staticmethod
     def forward(ctx: Any,
-                group: dist.ProcessGroup,
+                group: torch.distributed.ProcessGroup,
```
Comment (awan-10) [inline deepspeed/moe/sharded_moe.py:93]: add a todo comment here.

File: deepspeed/runtime/comm/coalesced_collectives.py [hunk 4]
```diff
@@ -6,36 +6,21 @@
 
 import torch
 from torch import Tensor
-import torch.distributed
+import deepspeed.comm as dist
+# TODO QUENTIN: Add process group class to comm
```
Comment (awan-10) [inline deepspeed/runtime/comm/coalesced_collectives.py:10]: cleanup

File: deepspeed/runtime/comm/coalesced_collectives.py [hunk 5]
```diff
@@ -6,36 +6,21 @@
...
-        input_tensor_lst = list(
-            torch.chunk(input_tensor,
-                        torch.distributed.get_world_size(group)))
-        instrument_w_nvtx(torch.distributed.reduce_scatter)(
-            output_tensor,
-            input_tensor_lst,
-            group=group,
-        )
+
+def torch_reduce_scatter_fn(input_tensor: Tensor, output_tensor: Tensor, group):
```
Comment (awan-10) [inline deepspeed/runtime/comm/coalesced_collectives.py:20]: either use _ or zero_ in the name

File: deepspeed/runtime/engine.py [hunk 6]
```diff
@@ -212,6 +215,11 @@ def __init__(
         self.moe_layers = []
         self._step_applied = False
         self._global_grad_norm = None
+        # TODO: Use a new ds_config flag for it. Hardcode for now
```
Comment (awan-10) [inline deepspeed/runtime/engine.py:218]: cleanup

File: deepspeed/runtime/engine.py [hunk 7]
```diff
@@ -224,16 +232,21 @@ def __init__(
...
-
-        if dist_init_required is False:
-            assert (
-                dist.is_initialized() is True
-            ), "Torch distributed not initialized. Please set dist_init_required to True or initialize before calling deepspeed.initialize()"
+        from deepspeed.comm import supported_torch_version
+        if supported_torch_version:
+            dist.init_distributed(dist_backend=self.dist_backend,
+                                  dist_init_required=dist_init_required)
         else:
```
Comment (awan-10) [inline deepspeed/runtime/engine.py:239]: Add a comment that this is for torch 1.2 only

File: deepspeed/runtime/pipe/module.py [hunk 8]
```diff
@@ -130,6 +130,8 @@ def forward(self, inputs):
         """
 
         super().__init__()
+        if not dist.is_initialized():
```
Comment (awan-10) [inline deepspeed/runtime/pipe/module.py:133]: @Quentin-Anthony, please investigate why this is needed? Revert it if its not needed.

File: deepspeed/runtime/zero/partition_parameters.py [hunk 9]
```diff
@@ -42,40 +42,16 @@
...
-        output_tensors = list(
-            torch.chunk(output_tensor,
-                        torch.distributed.get_world_size(group)))
-        return instrument_w_nvtx(torch.distributed.all_gather)(
-            output_tensors,
-            input_tensor,
-            group=group,
-            async_op=True,
-        )
+def dist_allgather_fn(input_tensor: Tensor, output_tensor: Tensor, group):
```
Comment (awan-10) [inline deepspeed/runtime/zero/partition_parameters.py:46]: _ in the name

File: op_builder/comm.py [hunk 10]
```diff
@@ -0,0 +1,35 @@
+import torch
```
Comment (awan-10) [inline op_builder/comm.py:1]: mark them as experimental

File: tests/comm/test.py [hunk 11]
```diff
@@ -0,0 +1,47 @@
+import torch
```
Comment (awan-10) [inline tests/comm/test.py:1]: delete it.

**Review by awan-10 — COMMENTED** [review]
I reviewed this PR with Quentin. He will take care of minor comments. @jeffra, please review this one.

File: tests/comm/test.py (hunk 11 above)
Comment (Quentin-Anthony) [reply tests/comm/test.py]: Done

File: op_builder/comm.py (hunk 10 above)
Comment (Quentin-Anthony) [reply op_builder/comm.py:3]: Added comment

File: deepspeed/runtime/zero/partition_parameters.py (hunk 9 above)
Comment (Quentin-Anthony) [reply deepspeed/runtime/zero/partition_parameters.py]: Done

File: deepspeed/runtime/engine.py (hunk 7 above)
Comment (Quentin-Anthony) [reply deepspeed/runtime/engine.py:239]: Added comment

File: deepspeed/runtime/engine.py (hunk 6 above)
Comment (Quentin-Anthony) [reply deepspeed/runtime/engine.py]: Done

File: deepspeed/runtime/comm/coalesced_collectives.py (hunk 5 above)
Comment (Quentin-Anthony) [reply deepspeed/runtime/comm/coalesced_collectives.py]: Done

File: deepspeed/runtime/comm/coalesced_collectives.py (hunk 4 above)
Comment (Quentin-Anthony) [reply deepspeed/runtime/comm/coalesced_collectives.py]: Done

File: deepspeed/moe/sharded_moe.py (hunk 3 above)
Comment (Quentin-Anthony) [reply deepspeed/moe/sharded_moe.py]: Done

File: deepspeed/runtime/pipe/module.py (hunk 8 above)
Comment (Quentin-Anthony) [reply deepspeed/runtime/pipe/module.py]: It was left over from your pipeline-parallel debugging. I removed it.

File: deepspeed/comm/backend.py (hunk 2 above)
Comment (Quentin-Anthony) [reply deepspeed/comm/backend.py]: Done. Please take a look.

File: deepspeed/__init__.py (hunk 1 above)
Comment (Quentin-Anthony) [reply deepspeed/__init__.py]: I reverted this. Our changes were unnecessary and caused errors.

**Comment by jeffra** [comment]
@SeanNaren can you take a look at the lightning test failure? We're seeing this on other PRs and in `master` right now as well. It appears to be a protobuf issue, have you seen this on your side before?

**Comment by Quentin-Anthony** [comment]
@jeffra -- This has been tested with HelloDeepSpeed, cifar (ZeRO 1,2,3 and MoE), and a modified [Megatron-Deepspeed](https://github.com/microsoft/Megatron-DeepSpeed/tree/ds_comms) (changed to use the new DeepSpeed comms API) with the following configurations and results:

https://gist.github.com/Quentin-Anthony/ed2082202b79451a15e946db174b4dbc

File: op_builder/comm.py [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+from .builder import CUDAOpBuilder
+
+MV2 = f'/opt/mvapich2/gdr/2.3.6/no-mpittool/no-openacc/cuda11.3/mofed5.4/mpirun/gnu9.3.0'
+OMPI = f'/usr/local/mpi'
+
+MPI_INCL = OMPI + '/include'
+MPI_LIB = '-L' + OMPI + '/lib -lmpi'
+
+CUDA = '/usr/local/cuda'
```
Comment (jeffra) [inline op_builder/comm.py:12]: We should pull the CUDA path from torch like this: https://github.com/microsoft/DeepSpeed/blob/6719b46bd8792755cfc75eaa448bf7fa65fd99eb/op_builder/builder.py#L39-L40

You'll need to do this from within the class below, so it's delayed to when the op is built.

**Comment by Quentin-Anthony** [comment]
@jeffra and @tjruwase -- Any further comments?

**Review by jeffra — APPROVED** [review]

File: deepspeed/comm/comm.py [hunk 13]
```diff
@@ -0,0 +1,541 @@
...
+
+    # TODO: load other ops. Clients including deepspeed itself should use deepspeed.comm to import
+    # any communication related primitives from this package.
+    # use hasattr(deepspeed.csrc.ops, "_comm") or something
+    return True
+
+
+def set_backend(backend):
+    if not use_ds_backend:
+        utils.logger.warn(
```
Comment (tjruwase) [inline deepspeed/comm/comm.py:122]: Can we fail on this? A common complaint is that deepspeed generates too many log messages that overwhelm users. I think if there is a no reasonable progress for users in this case, then we should fail as users are likely to miss this warning message.

File: deepspeed/comm/comm.py [hunk 14]
```diff
@@ -0,0 +1,541 @@
...
+
+
+# UNUSED: Future helper function to initialize DS backends
+def init_deepspeed_backend(ds_backend):
+    global cdb
+    global nccl_backend
+    global mpi_backend
+    global use_ds_backend
+
+    if ds_backend == 'nccl':
```
Comment (tjruwase) [inline deepspeed/comm/comm.py:81]: Please replace hardcoded strings like 'nccl' with vars.

File: deepspeed/comm/comm.py [hunk 15]
```diff
@@ -0,0 +1,541 @@
...
+            "DeepSpeed communication backend is required. Please use deepspeed.comm.init_process_group(backend, use_deepspeed=True) to use this functionality"
+        )
+        return
+
+    global cdb
+    global nccl_backend
+    global mpi_backend
+
+    try:
+        if backend_name == 'nccl':
```
Comment (tjruwase) [inline deepspeed/comm/comm.py:132]: Is `backend_name` a global var? I can't find the definition. Thanks.

File: deepspeed/comm/comm.py [hunk 16]
```diff
@@ -0,0 +1,541 @@
...
+
+
+def set_backend(backend):
+    if not use_ds_backend:
+        utils.logger.warn(
+            "DeepSpeed communication backend is required. Please use deepspeed.comm.init_process_group(backend, use_deepspeed=True) to use this functionality"
+        )
+        return
+
+    global cdb
```
Comment (tjruwase) [inline deepspeed/comm/comm.py:127]: What does cdb stand for?

File: deepspeed/comm/comm.py [hunk 17]
```diff
@@ -0,0 +1,541 @@
...
+                mpi_discovery(distributed_port=distributed_port, verbose=verbose)
+
+        if cdb is None or not cdb.is_initialized():
+            if verbose and int(os.getenv('RANK', '0')) == 0:
+                utils.logger.info(
+                    "Initializing torch distributed with backend: {}".format(
+                        dist_backend))
+            assert isinstance(timeout, timedelta)
+
+            if cdb is not None and cdb.is_initialized():
```
Comment (tjruwase) [inline deepspeed/comm/comm.py:421]: Condition 421 seems the opposite of 414. Can you please double check?

File: deepspeed/comm/comm.py [hunk 18]
```diff
@@ -0,0 +1,541 @@
...
+                if int(os.getenv('RANK', '0')) == 0:
+                    utils.logger.info('Distributed backend already initialized')
```
Comment (tjruwase) [inline deepspeed/comm/comm.py:423]: Reading message 423 after 416 could be confusing. Can we simplify the logic here so that only one of these shows up in the log?

File: deepspeed/runtime/zero/stage_1_and_2.py [hunk 19]
```diff
@@ -517,7 +516,14 @@ def is_moe_group(self, group):
         return 'moe' in group and group['moe']
 
     def _configure_moe_settings(self):
-        assert self.contiguous_gradients, "Contiguous Gradients in ZeRO Stage 2 must be set to True for MoE. Other code paths are not tested with MoE"
+        # if we're using ZeRO stage 2, ensure contiguous gradients are used
```
Comment (tjruwase) [inline deepspeed/runtime/zero/stage_1_and_2.py:519]: This difference seems unrelated to communication backend. Please double check.

File: deepspeed/comm/comm.py [hunk 20]
```diff
@@ -0,0 +1,541 @@
...
+
+from deepspeed.comm.backend import Backend
+from deepspeed.comm.torch import TorchBackend
+
+import deepspeed.utils as utils
+from datetime import timedelta
+
+# Current deepspeed.comm backend (cdb) global object for simple access by client code
+use_ds_backend = False
+cdb = None
```
Comment (tjruwase) [inline deepspeed/comm/comm.py:56]: What does cdb mean?

Comment (tjruwase) [reply deepspeed/comm/comm.py:56]: deepspeed comm backend => dcb?

**Comment by rocm-mici** [comment]
Can one of the admins verify this patch?

File: deepspeed/runtime/zero/stage_1_and_2.py (hunk 19 above)
Comment (Quentin-Anthony) [reply deepspeed/runtime/zero/stage_1_and_2.py:519]: This arose from my Megatron-DeepSpeed testing. This assertion was failing MoE with ZeRO-1, which is a supported mode. I discussed this with @awan-10 and we agreed on this fix.

Would you like me to create a separate PR for this or can it be bundled with this one as-is?

File: deepspeed/comm/comm.py (hunk 20 above)
Comment (Quentin-Anthony) [reply deepspeed/comm/comm.py:56]: It stands for "current deepspeed.comm backend": https://github.com/microsoft/DeepSpeed/blob/staging-comms-next/deepspeed/comm/comm.py#L54

I'm OK with a cdb => dcb refactor if @awan-10 is. I agree that it's a more natural var name.

File: deepspeed/comm/comm.py (hunk 14 above)
Comment (Quentin-Anthony) [reply deepspeed/comm/comm.py:81]: Would you prefer a global variable name `deepspeed/comm/comm.py` (e.g. `NCCL_BACKEND`) or a constant of the same name in `deepspeed/constants.py`?

File: deepspeed/comm/comm.py (hunk 13 above)
Comment (Quentin-Anthony) [reply deepspeed/comm/comm.py]: Done

File: deepspeed/comm/comm.py (hunk 17 above)
Comment (Quentin-Anthony) [reply deepspeed/comm/comm.py]: Good catch! I've cleaned up this logic. Take a look at the changes and let me know if you'd like further updates.

File: deepspeed/comm/comm.py (hunk 18 above)
Comment (Quentin-Anthony) [reply deepspeed/comm/comm.py]: I've cleaned this up along with the comment above. Take a look and let me know if you'd like further updates.

File: deepspeed/comm/comm.py (hunk 16 above)
Comment (Quentin-Anthony) [reply deepspeed/comm/comm.py:129]: It stands for "current deepspeed.comm backend": https://github.com/microsoft/DeepSpeed/blob/staging-comms-next/deepspeed/comm/comm.py#L54

I'm OK with a cdb => dcb refactor if @awan-10 is. I agree that it's a more natural var name.

File: deepspeed/runtime/zero/stage_1_and_2.py (hunk 19 above)
Comment (tjruwase) [reply deepspeed/runtime/zero/stage_1_and_2.py:519]: Yes, please. A separate PR will help track things in future.

File: deepspeed/comm/comm.py (hunk 20 above)
Comment (tjruwase) [reply deepspeed/comm/comm.py:56]: Either way is fine with me. It was not obvious to me.

File: deepspeed/comm/comm.py (hunk 14 above)
Comment (jeffra) [reply deepspeed/comm/comm.py:81]: Perhaps `deepspeed/comm/constants.py` would be a good place for these

Comment (Quentin-Anthony) [reply deepspeed/comm/comm.py]: > Perhaps `deepspeed/comm/constants.py` would be a good place for these

I've added a `deepspeed/comm/constants.py` file containing all string constant assignments I could find.

File: deepspeed/runtime/zero/stage_1_and_2.py (hunk 19 above)
Comment (Quentin-Anthony) [reply deepspeed/runtime/zero/stage_1_and_2.py]: I reverted this commit and created a separate PR here: https://github.com/microsoft/DeepSpeed/pull/2007

**Review by tjruwase — APPROVED** [review]
